docs/Server-Methods.md

```diff
@@ -138,3 +138,14 @@ Set the schema compiler for all routes [here](https://github.com/fastify/fastify
 #### setErrorHandler
 
 `fastify.setErrorHandler(handler(error, reply))`: set a function that will be called whenever an error happens. The handler is fully encapsulated, so different plugins can set different error handlers, *async await* is supported as well.
+
+This handler can be called if an error occurs after headers are already sent. If this happens you will not be able to send a response, so you can avoid any unnecessary sending logic by checking the `res.headersSent` property.
+
+```js
+fastify.setErrorHandler(function (error, reply) {
+  if (reply.res.headersSent) {
```

Comment:
Should we perform this check internally and not call the custom error handler at all?

Reply:
That's what this change is all about. The custom error handler *should* be called so that users have a chance to log the error in their own way (not everyone has Fastify's logger enabled).

Reply:
Ops, I misunderstood then.
I don't know if I'm really ok with this change, this could be really annoying.
Since there is the chance to use an external logger I don't see why a user should write some complex code to integrate an external logger that is not integrated within the framework.
I'd like to have the opinion of the other on this point.

Reply:
I'm not ok with this change either.
Given the logger is pluggable, I don't see much of a reason to make the user aware of this for only logging purposes.

Reply:
It's not really about integrating a logger using Fastify's `logger` option. It's more that not everyone agrees with the way that things are logged. Maybe I want some [400 errors to be logged](https://github.com/fastify/fastify/blob/629c16c9c9d09109a264f8868f7f0844907bb5a3/lib/reply.js#L198-L199) as `warn` instead of `info`, or to log more information in addition to `res` and `err`. Maybe I want to log stream errors differently than other errors.

Some people switch to Fastify from something else and already have specific log messages and analytics for those messages. Not giving users the chance to log things in their own way (or at all) isn't very nice. 

I agree this change seems wrong if you see the custom error handler as a way to send a response like a route handler, but I see the custom error handler as a way to handle the error in my own way.

Reply:
A solution of this might be to make `res.send()` idempotent if the headers are already been sent. However, this might cause more damage than what we are trying to solve, because it can hide some bugs.

The error handler is for errors that can be acted upon. That's why it gets a `reply` object. Errors that cannot get acted upon should only be logged. BTW, I'm fine if we log with warn level.

If you want to handle this case differently, you can do a lot of other things. Sorry.